include/extractor/guidance/graph_hopper.hpp

```diff
@@ -0,0 +1,143 @@
+#ifndef OSRM_EXTRACTOR_GUIDANCE_GRAPH_HOPPER
+#define OSRM_EXTRACTOR_GUIDANCE_GRAPH_HOPPER
+
+#include "extractor/guidance/constants.hpp"
+#include "extractor/guidance/intersection_generator.hpp"
+#include "extractor/guidance/toolkit.hpp"
+#include "extractor/guidance/toolkit.hpp"
+#include "util/coordinate.hpp"
+#include "util/node_based_graph.hpp"
+#include "util/typedefs.hpp"
+
+#include <boost/assert.hpp>
+#include <boost/optional.hpp>
+#include <utility>
+
+namespace osrm
+{
+namespace extractor
+{
+namespace guidance
+{
+
+// forward declaration to allow interaction between the intersection generator and the graph hopper
+class IntersectionGenerator;
+
+/*
+ * The graph hopper is a utility that lets you find certain intersections with a node based graph,
+ * accumulating information along the way
+ */
+class GraphHopper
```

Comment:
Not sure you really want to call it that - there is a routing engine out there with that name, might be confusing for people stumbling over it.